File: app/src/main/java/com/moggot/findmycarlocation/presentation/common/BasePresenter.java
package com.moggot.findmycarlocation.presentation.common;

import android.support.annotation.NonNull;

import io.reactivex.disposables.CompositeDisposable;
import io.reactivex.disposables.Disposable;

public class BasePresenter<T extends BaseView> {

    private final CompositeDisposable disposable = new CompositeDisposable();
    private T view;

    protected void unSubscribeOnDetach(Disposable... disposables) {
        disposable.addAll(disposables);
    }

    public void onAttach(@NonNull T view) {
        this.view = view;
    }

    public void onDetach() {
        disposable.clear();
        this.view = null;
    }

    protected T getView() {
        return view;
    }
}

File: app/src/main/java/com/moggot/findmycarlocation/base/BasePresenter.kt
package com.moggot.findmycarlocation.base

import io.reactivex.disposables.CompositeDisposable
import io.reactivex.disposables.Disposable

abstract class BasePresenter<V : BaseView> {

    protected var view: V? = null
    protected val disposable = CompositeDisposable()

    open fun unSubscribeOnDetach(vararg disposables: Disposable) {
        disposable.addAll(*disposables)
    }

    fun attachView(view: V) {
        this.view = view
        onViewAttached(view)
    }

    open fun onViewAttached(view: V) {}

    fun view(action: V.() -> Unit) {
        view?.run(action)
    }

    fun detachView() {
        disposable.clear()
        view = null
    }
}